File: mdix-java/src/main/test/java/com/midmanstudio/dixscript/QueryTest.java
package com.midmanstudio.dixscript;

import org.junit.jupiter.api.*;
import static org.assertj.core.api.Assertions.*;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Integration tests for {@link Database#query}, {@link Database#queryMany}, and {@link MdixQuery}.
 * Requires the native lib to be on java.library.path (set by build.gradle.kts).
 */
class QueryTest {

    private static final String ENEMIES_SRC =
        "@DATA( " +
        "  world = \"overworld\" " +
        "  tags:: \"alpha\", \"beta\", \"alpha\", \"gamma\" " +
        "  enemies:: " +
        "    { name = \"Goblin\", health = 50, aiType = \"AGGRESSIVE\" }, " +
        "    { name = \"Orc\", health = 100, aiType = \"AGGRESSIVE\" }, " +
        "    { name = \"Dragon\", health = 1000, aiType = \"BOSS\" }, " +
        "    { name = \"Slime\", health = 20, aiType = \"PASSIVE\" } " +
        "  levels:: " +
        "    { id = 1, enemies:: { name = \"Rat\", health = 5, aiType = \"PASSIVE\" } }, " +
        "    { id = 2, enemies:: { name = \"Bat\", health = 8, aiType = \"PASSIVE\" } } " +
        ")";

    private Database db;

    @BeforeEach
    void setUp() {
        db = DixScript.loadStr(ENEMIES_SRC);
    }

    @AfterEach
    void tearDown() {
        db.close();
    }

    // ── query(path) ───────────────────────────────────────────────────────────

    @Test void query_returnsAllElements() {
        assertThat(db.query("enemies").count()).isEqualTo(4);
    }

    @Test void query_where_filtersByField() {
        List<String> aggressive = db.query("enemies")
            .where_(e -> "AGGRESSIVE".equals(e.field("aiType").asString()))
            .select(e -> e.field("name").asString());
        assertThat(aggressive).containsExactly("Goblin", "Orc");
    }

    @Test void query_whereFieldEquals_shorthand() {
        assertThat(db.query("enemies").whereFieldEquals("aiType", MdixValue.ofString("BOSS")).count())
            .isEqualTo(1);
    }

    @Test void query_select_projectsField() {
        List<String> names = db.query("enemies").selectField("name").stream()
            .map(MdixValue::asString)
            .collect(Collectors.toList());
        assertThat(names).containsExactly("Goblin", "Orc", "Dragon", "Slime");
    }

    @Test void query_orderBy_sortsAscending() {
        List<Long> healths = db.query("enemies")
            .orderBy(e -> e.field("health").asLong())
            .select(e -> e.field("health").asLong());
        assertThat(healths).containsExactly(20L, 50L, 100L, 1000L);
    }

    @Test void query_orderByDescending_sortsDescending() {
        MdixValue first = db.query("enemies").orderByDescending(e -> e.field("health").asLong()).first();
        assertThat(first.field("name").asString()).isEqualTo("Dragon");
    }

    @Test void query_take_limitsResults() {
        assertThat(db.query("enemies").take(2).count()).isEqualTo(2);
    }

    @Test void query_skip_dropsLeadingResults() {
        assertThat(db.query("enemies").skip(3).count()).isEqualTo(1);
    }

    @Test void query_any_true() {
        assertThat(db.query("enemies").any(e -> "BOSS".equals(e.field("aiType").asString()))).isTrue();
    }

    @Test void query_all_false() {
        assertThat(db.query("enemies").all(e -> "AGGRESSIVE".equals(e.field("aiType").asString()))).isFalse();
    }

    @Test void query_count_withFilter() {
        long aggressiveCount = db.query("enemies")
            .where_(e -> "AGGRESSIVE".equals(e.field("aiType").asString()))
            .count();
        assertThat(aggressiveCount).isEqualTo(2);
    }

    @Test void query_isEmpty_falseWhenPopulated() {
        assertThat(db.query("enemies").isEmpty()).isFalse();
    }

    @Test void query_first_and_last() {
        assertThat(db.query("enemies").first().field("name").asString()).isEqualTo("Goblin");
        assertThat(db.query("enemies").last().field("name").asString()).isEqualTo("Slime");
    }

    @Test void query_nth_returnsElementAtIndex() {
        assertThat(db.query("enemies").nth(2).field("name").asString()).isEqualTo("Dragon");
    }

    @Test void query_nth_outOfRange_returnsNull() {
        assertThat(db.query("enemies").nth(99)).isNull();
    }

    @Test void query_sumInt_sumsHealth() {
        assertThat(db.query("enemies").select(e -> e.field("health").asLong())
            .stream().mapToLong(Long::longValue).sum()).isEqualTo(1170L);
    }

    @Test void query_groupBy_groupsByField() {
        Map<String, List<MdixValue>> byAi = db.query("enemies").groupBy(e -> e.field("aiType").asString());
        assertThat(byAi.get("AGGRESSIVE")).hasSize(2);
        assertThat(byAi.get("BOSS")).hasSize(1);
        assertThat(byAi.get("PASSIVE")).hasSize(1);
    }

    @Test void query_distinct_removesDuplicateScalars() {
        assertThat(db.query("tags").count()).isEqualTo(4);
        assertThat(db.query("tags").distinct().count()).isEqualTo(3);
    }

    // ── queryMany(pattern) ───────────────────────────────────────────────────

    @Test void queryMany_nonMatchingPattern_returnsEmpty() {
        assertThat(db.queryMany("no.such.path.*").isEmpty()).isTrue();
    }

    @Test void queryMany_globPattern_doesNotThrow() {
        // Exact glob-matcher semantics live in dixscript::Runtime::DixData::select_many;
        // this is a smoke test that the native call round-trips cleanly end to end.
        assertThatCode(() -> db.queryMany("levels.*.enemies")).doesNotThrowAnyException();
    }

    // ── stream() escape hatch ────────────────────────────────────────────────

    @Test void query_stream_dropsToJavaStream() {
        long count = db.query("enemies").stream()
            .filter(e -> e.field("health").asLong() != null && e.field("health").asLong() > 30)
            .count();
        assertThat(count).isEqualTo(3);
    }
}

File: mdix-php/tests/QueryTest.php
<?php
declare(strict_types=1);

namespace MidManStudio\Mdix\Tests;

use MidManStudio\Mdix\MdixDatabase;
use PHPUnit\Framework\TestCase;

/**
 * Integration tests for MdixDatabase::query()/queryMany() and MdixQuery.
 * Requires the native lib to be on MDIX_LIB_PATH or in mdix-php/lib/.
 */
final class QueryTest extends TestCase
{
    private const ENEMIES_SRC = <<<'MDIX'
@DATA(
  world = "overworld"
  tags:: "alpha", "beta", "alpha", "gamma"
  enemies::
    { name = "Goblin", health = 50,   aiType = "AGGRESSIVE" },
    { name = "Orc",     health = 100, aiType = "AGGRESSIVE" },
    { name = "Dragon",  health = 1000, aiType = "BOSS" },
    { name = "Slime",   health = 20,  aiType = "PASSIVE" }
  levels::
    { id = 1, enemies:: { name = "Rat", health = 5, aiType = "PASSIVE" } },
    { id = 2, enemies:: { name = "Bat", health = 8, aiType = "PASSIVE" } }
)
MDIX;

    private MdixDatabase $db;

    protected function setUp(): void
    {
        $this->db = MdixDatabase::loadStr(self::ENEMIES_SRC);
    }

    protected function tearDown(): void
    {
        $this->db->close();
    }

    public function testQueryReturnsAllElements(): void
    {
        self::assertSame(4, $this->db->query('enemies')->count());
    }

    public function testWhereFiltersByField(): void
    {
        $aggressive = $this->db->query('enemies')
            ->where(fn (array $e): bool => $e['aiType'] === 'AGGRESSIVE')
            ->select(fn (array $e): string => $e['name']);

        self::assertSame(['Goblin', 'Orc'], $aggressive);
    }

    public function testWhereFieldEqualsShorthand(): void
    {
        self::assertSame(1, $this->db->query('enemies')->whereFieldEquals('aiType', 'BOSS')->count());
    }

    public function testSelectProjectsField(): void
    {
        self::assertSame(
            ['Goblin', 'Orc', 'Dragon', 'Slime'],
            $this->db->query('enemies')->selectField('name'),
        );
    }

    public function testOrderBySortsAscending(): void
    {
        $healths = $this->db->query('enemies')
            ->orderBy(fn (array $e): int => $e['health'])
            ->selectField('health');

        self::assertSame([20, 50, 100, 1000], $healths);
    }

    public function testOrderByDescendingSortsDescending(): void
    {
        $first = $this->db->query('enemies')->orderByDescending(fn (array $e): int => $e['health'])->first();
        self::assertSame('Dragon', $first['name']);
    }

    public function testTakeLimitsResults(): void
    {
        self::assertSame(2, $this->db->query('enemies')->take(2)->count());
    }

    public function testSkipDropsLeadingResults(): void
    {
        self::assertSame(1, $this->db->query('enemies')->skip(3)->count());
    }

    public function testAnyTrue(): void
    {
        self::assertTrue($this->db->query('enemies')->any(fn (array $e): bool => $e['aiType'] === 'BOSS'));
    }

    public function testAllFalse(): void
    {
        self::assertFalse($this->db->query('enemies')->all(fn (array $e): bool => $e['aiType'] === 'AGGRESSIVE'));
    }

    public function testCountWithFilter(): void
    {
        $count = $this->db->query('enemies')->where(fn (array $e): bool => $e['aiType'] === 'AGGRESSIVE')->count();
        self::assertSame(2, $count);
    }

    public function testIsEmptyFalseWhenPopulated(): void
    {
        self::assertFalse($this->db->query('enemies')->isEmpty());
    }

    public function testFirstAndLast(): void
    {
        self::assertSame('Goblin', $this->db->query('enemies')->first()['name']);
        self::assertSame('Slime', $this->db->query('enemies')->last()['name']);
    }

    public function testNthReturnsElementAtIndex(): void
    {
        self::assertSame('Dragon', $this->db->query('enemies')->nth(2)['name']);
    }

    public function testNthOutOfRangeReturnsNull(): void
    {
        self::assertNull($this->db->query('enemies')->nth(99));
    }

    public function testSumIntSumsHealth(): void
    {
        $sum = \array_sum($this->db->query('enemies')->selectField('health'));
        self::assertSame(1170, $sum);
    }

    public function testGroupByGroupsByField(): void
    {
        $byAi = $this->db->query('enemies')->groupBy(fn (array $e): string => $e['aiType']);
        self::assertCount(2, $byAi['AGGRESSIVE']);
        self::assertCount(1, $byAi['BOSS']);
        self::assertCount(1, $byAi['PASSIVE']);
    }

    public function testDistinctRemovesDuplicateScalars(): void
    {
        self::assertSame(4, $this->db->query('tags')->count());
        self::assertSame(3, $this->db->query('tags')->distinct()->count());
    }

    public function testQueryManyNonMatchingPatternReturnsEmpty(): void
    {
        self::assertTrue($this->db->queryMany('no.such.path.*')->isEmpty());
    }

    public function testQueryManyGlobPatternDoesNotThrow(): void
    {
        // Exact glob-matcher semantics live in dixscript::Runtime::DixData::select_many;
        // this is a smoke test that the native call round-trips cleanly end to end.
        $this->db->queryMany('levels.*.enemies');
        $this->addToAssertionCount(1);
    }

    public function testToArrayDropsToPlainPhpArray(): void
    {
        $high = \array_filter(
            $this->db->query('enemies')->toArray(),
            fn (array $e): bool => $e['health'] > 30,
        );
        self::assertCount(3, $high);
    }
}
